doc/read-the-docs-site/adr/0005-plutus-tx-compiler-and-library.md

```diff
@@ -0,0 +1,209 @@
+# ADR 5: Plutus Tx's Evaluation Strategy, Compiler and Standard Library
+
+Date: 2023-04
+
+## Authors
+
+Ziyang Liu <ziyang.liu@iohk.io>
+
+## Status
+
+Proposed
+
+## Background
+
+Plutus Tx is a language for writing Plutus scripts.
+It is a subset of Haskell and can be compiled to Untyped Plutus Core (hereafter referred to as PLC) by the Plutus Tx compiler, a GHC plugin.
+
+Haskell uses lazy evaluation (call by need) by default, while PLC is strict (call by value).
+This begs the question: what is the operational semantics of Plutus Tx?
+This is an important question, the answer of which not only dictates how users of Plutus Tx should reason about and optimize their programs, but also has major implications on the design and implementation of the Plutus Tx compiler and libraries.
+
+Unfortunately, this question is currently difficult to answer.
+One thing is clear: Plutus Tx certainly doesn't have lazy semantics: PLC's evaluator (the CEK machine) has no support for thunks and memoization.
+What is unclear is whether Plutus Tx is strict or non-strict.
+Indeed, it is sometimes strict and sometimes non-strict, and it is fairly random and difficult to predict:
+
+- Let-bindings in Plutus Tx are by default non-strict, but function applications are strict.
+  That is, `let x = rhs in body` isn't equivalent to `(\x -> body) rhs`.
+  This is distinct from most other languages, regardless of their evaluation strategies, in which these two expressions are equivalent.
+
+- The conditional expression `if b then t else f` is always non-strict in `t` and `f`, like in most strict languages.
+  But what about `foo b t f = if b then t else f`? In a strict language, `foo` would be strict, and in a non-strict language, it would be non-strict.
+  But in Plutus Tx, due to how its compiler is currently implemented, whether `t` and `f` are evaluated strictly or not depends on whether GHC inlines `foo`.
+  Although we turn off GHC's inliner in the Plutus Tx compiler, GHC may still decide to unconditionally inline `foo` if it is used only once.
+  In general, it is not easy to predict whether GHC will inline something or not.
```

Comment:
It's easy if you turn off all GHC optimization. RV didn't want any inlining whatsoever and turning off all GHC optimization with `-fmax-simplifier-iterations=0` worked for them.

Reply:
Nope, not at all.